cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableRegistry.java

```diff
@@ -0,0 +1,155 @@
+    this.entityRegistrySpec = tableDefinition.configure(ENTITY_REGISTRY, DatasetProperties.EMPTY);
+    this.transactional = Transactions.createTransactionalWithRetry(
+      NoSqlTransactionRunner.createTransactional(txClient, this::getRegistryTable),
+      RetryStrategies.retryOnConflict(20, 100));
+  }
+
+  @Override
+  public void initialize() throws IOException {
+    DatasetAdmin admin = tableDefinition.getAdmin(SYSTEM_CONTEXT, entityRegistrySpec, null);
+    if (!admin.exists()) {
+      LOG.info("Creating dataset table {} in namespace {}", entityRegistrySpec.getName(), NamespaceId.SYSTEM);
+      admin.create();
+    }
+  }
+
+  @Override
+  public void registerSpecification(StructuredTableSpecification specification) throws TableAlreadyExistsException {
+    LOG.debug("Registering table specification {}", specification);
+    StructuredTableId tableId = specification.getTableId();
+    Transactionals.execute(
+      transactional,
+      context -> {
+        Table table = context.getDataset(REGISTRY_DATASET_ID.getDataset());
+        byte[] serialized = table.get(getRowKeyBytes(tableId), SCHEMA_COL_BYTES);
+        if (serialized != null) {
+          throw new TableAlreadyExistsException(tableId);
+        }
+        serialized = Bytes.toBytes(GSON.toJson(specification));
+        table.put(getRowKeyBytes(tableId), SCHEMA_COL_BYTES, serialized);
+      }, TableAlreadyExistsException.class);
+  }
+
+  @Nullable
+  @Override
+  public StructuredTableSpecification getSpecification(StructuredTableId tableId) {
+    return Transactionals.execute(
+      transactional,
+      context -> {
+        Table table = context.getDataset(REGISTRY_DATASET_ID.getDataset());
+        byte[] serialized = table.get(getRowKeyBytes(tableId), SCHEMA_COL_BYTES);
+        if (serialized == null) {
+          return null;
+        }
+        return GSON.fromJson(Bytes.toString(serialized), StructuredTableSpecification.class);
+      }
+    );
+  }
+
+  @Override
+  public void removeSpecification(StructuredTableId tableId) {
+    LOG.debug("Removing table specification for table {}", tableId);
+    Transactionals.execute(
+      transactional,
+      context -> {
+        Table table = context.getDataset(REGISTRY_DATASET_ID.getDataset());
+        table.delete(getRowKeyBytes(tableId));
+      });
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return Transactionals.execute(
+      transactional,
+      context -> {
+        Table table = context.getDataset(REGISTRY_DATASET_ID.getDataset());
+        try (Scanner scanner = table.scan(TABLE_ROWKEY_PREFIX, Bytes.stopKeyForPrefix(TABLE_ROWKEY_PREFIX))) {
+          return scanner.next() == null;
+        }
+      }
+    );
+  }
+
+  private static byte[] getRowKeyBytes(StructuredTableId tableId) {
+   return Bytes.concat(TABLE_ROWKEY_PREFIX, Bytes.toBytes(tableId.getName()));
+  }
+
+  private <T extends Dataset> T getRegistryTable(String name) throws IOException {
+    if (ENTITY_REGISTRY.equals(name)) {
```

Comment:
Same here, why need to take the name if it has to be `ENTITY_REGISTRY`?

Reply:
Moved the check to `TableDatasetSupplier` instance